File: code/ex994.java
class Solution {
    public int orangesRotting(int[][] grid) {
        int m = grid.length, n = grid[0].length;
        
        Queue<int[]> queue = new LinkedList<>();
        boolean[][] visited = new boolean[m][n];
        int left = 0;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (grid[i][j] == 2) {
                    visited[i][j] = true;
                    queue.offer(new int[]{i, j});
                }
                else if (grid[i][j] == 1)    left++;
            }
        }
        if (left == 0)    return 0;
        
        // Go BFS
        int minute = 0;
        int[][] dir = new int[][]{{0, 1}, {0, -1}, {1, 0}, {-1, 0}};
        while (!queue.isEmpty()) {
            for (int z = queue.size(); z > 0; z--) {
                int[] out = queue.poll();
                int i = out[0], j = out[1];
                
                for (int d = 0; d < 4; d++) {
                    int ni = i + dir[d][0], nj = j + dir[d][1];
                    if (ni < 0 || nj < 0 || ni >= m || nj >= n || visited[ni][nj])    continue;
                    
                    if (grid[ni][nj] == 1) {
                        grid[ni][nj] = 2;
                        visited[ni][nj] = true;
                        if (--left == 0)    return minute + 1;
                        queue.offer(new int[]{ni, nj});
                    }
                }
            }
            minute++;
        }
        
        return -1;
    }
}

File: code/ex994.cpp
class Solution {
public:
    int orangesRotting(vector<vector<int>>& matrix) {
        int m = matrix.size(), n = matrix[0].size();
		vector<vector<int>> dir = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};

		queue<vector<int>> que;
		int freshOranges = 0;
		for (int i = 0; i < m; i++)
			for (int j = 0; j < n; j++) {
				if (matrix[i][j] == 2)
					que.push({i, j});
                else if (matrix[i][j] == 1)
                    freshOranges++;
            }
        if (freshOranges == 0)    return 0;

		int minutes = 0;
		while (!que.empty()) {
			int size = que.size();
			for (int z = size; z > 0; z--) {
				auto out = que.front();    que.pop();
				int i = out[0], j = out[1];
				for (auto& d : dir) {
					int ni = i + d[0], nj = j + d[1];
					if (ni < 0 || nj < 0 || ni >= m || nj >= n || matrix[ni][nj] != 1)    continue;
					matrix[ni][nj] = 2;
					if (--freshOranges == 0)    return minutes + 1;
					que.push({ni, nj});
				}
			}
			minutes++;
		}

		return -1;
    }
};